subsys/bluetooth/host/settings.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <zephyr.h>
+#include <settings/settings.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_SETTINGS)
+#include "common/log.h"
+
+#include "hci_core.h"
+
+static int set(int argc, char **argv, char *val)
+{
+	int len;
+
+	BT_DBG("argc %d argv[0] %s val %s", argc, argv[0], val);
+
+	if (argc != 1) {
+		return -ENOENT;
+	}
+
+	if (!strcmp(argv[0], "id")) {
+		len = sizeof(bt_dev.id_addr);
+		settings_bytes_from_str(val, &bt_dev.id_addr, &len);
+		BT_DBG("ID Addr set to %s", bt_addr_le_str(&bt_dev.id_addr));
+		return 0;
+	}
+
+#if defined(CONFIG_BT_PRIVACY)
+	if (!strcmp(argv[0], "irk")) {
+		len = sizeof(bt_dev.irk);
+		settings_bytes_from_str(val, bt_dev.irk, &len);
+		BT_DBG("IRK set to %s", bt_hex(bt_dev.irk, 16));
+		return 0;
+	}
+#endif /* CONFIG_BT_PRIVACY */
+
+	return 0;
+}
+
+static void generate_static_addr(void)
+{
+	char buf[13];
+	char *str;
+
+	BT_DBG("Generating new static random address");
+
+	if (bt_addr_le_create_static(&bt_dev.id_addr)) {
+		BT_ERR("Failed to generate static addr");
+		return;
+	}
+
+	bt_set_static_addr();
+
+	BT_DBG("New ID Addr: %s", bt_addr_le_str(&bt_dev.id_addr));
+
+	str = settings_str_from_bytes(&bt_dev.id_addr, sizeof(bt_dev.id_addr),
+				      buf, sizeof(buf));
+	if (!str) {
+		BT_ERR("Unable to encode ID Addr as value");
+		return;
+	}
+
+	BT_DBG("Saving ID addr as value %s", str);
+	settings_save_one("bt/id", str);
+}
+
+#if defined(CONFIG_BT_PRIVACY)
+static void generate_irk(void)
+{
+	char buf[25];
+	char *str;
+
+	BT_DBG("Generating new IRK");
+
+	if (bt_rand(bt_dev.irk, sizeof(bt_dev.irk))) {
+		BT_ERR("Failed to generate IRK");
+		return;
+	}
+
+	BT_DBG("New local IRK: %s", bt_hex(bt_dev.irk, 16));
+
+	str = settings_str_from_bytes(bt_dev.irk, 16, buf, sizeof(buf));
+	if (!str) {
+		BT_ERR("Unable to encode IRK as value");
+		return;
+	}
+
+	BT_DBG("Saving IRK as value %s", str);
+	settings_save_one("bt/irk", str);
+}
+#endif /* CONFIG_BT_PRIVACY */
+
+static int commit(void)
+{
+	BT_DBG("");
+
+	if (!bt_addr_le_cmp(&bt_dev.id_addr, BT_ADDR_LE_ANY) ||
+	    !bt_addr_le_cmp(&bt_dev.id_addr, BT_ADDR_LE_NONE)) {
+		generate_static_addr();
+	}
+
+#if defined(CONFIG_BT_PRIVACY)
+	{
+		u8_t zero[16] = { 0 };
+
+		if (!memcmp(bt_dev.irk, zero, 16)) {
+			generate_irk();
+		}
+	}
+#endif /* CONFIG_BT_PRIVACY */
+
+	bt_dev_show_info();
+
+	return 0;
+}
+
+static struct settings_handler bt_settings = {
+	.name = "bt",
+	.h_set = set,
+	.h_commit = commit,
+};
+
+int bt_settings_init(void)
+{
+	int err;
+
+	BT_DBG("");
+
+	err = settings_subsys_init();
```

Comment:
Note about this: I'm not 100% sure if the settings subsystem ensures that this API can be called multiple times (potentially by multiple subsystems). At least parts of it were protected by an "is_inited" boolean, but not all, I think. This might be good to fix in settings. Alternatively, we could move the requirement of calling settings_subsys_init() to the application side, i.e. before calling bt_enable(). That would make it consistent with the requirement of calling settings_load() on the application side as well. Thoughts?

Reply:
Before doing that however, we should fix settings so that settings_register() either cleanly fails or cleanly succeeds if called before settings_subsys_init() has been called. Currently that's not that case since settings_register() will always succeed, however settings_subsys_init() will clear the list of handlers if it's called later. I'd propose using static initialization of the handlers list, so that the sys_slist_init() call in settings_store_init() is not needed. That way it doesn't matter in which order things happen, and we can always safely call settings_register().

That said, I think I might still prefer for bt_enable() to fail if the application has not called settings_subsys_init(), which means that settings_register() should return failure in that case. Any thoughts @nvlsianpu ?

Reply:
One more thought: if we move the settings handlers to a dedicated ROM linker section, the settings_register() API will go away. That would change things a bit. One of the reasons I did settings_load() outside of bt_enable() is that with mesh the mesh settings handlers would only be registered upon bt_mesh_init() (which is called after bt_enable). However, if we have a dedicated linker sections, all handlers are already there from the very beginning, making it safer to call settings_load() at any time (it might still not be completely safe though since there might still be run-time init functions that some subsystems need to get called first)

Reply:
For the record: I've just pushed a third patch to this PR, showing how I'd do a ROM linker section for the Bluetooth-internal submodule handlers. The solution for the settings handlers would look quite similar.

Reply:
Agree, but have in mind that fcb test `tests/subsys/settings/fcb/src/settings_test_fcb.c'` need to be aligned as settings_register() is called without initaialization proir (any test_config_insert* testcase)

Reply:
Agree with the dependency on fixing `settings_register` before we can actually safely consider the initialization path. So should we put this PR on hold while we send a PR to fix the settings subsystem itself?

Reply:
Think not necessarily hood on this - as can be implemented properly even now.